Kaylee/many to many test

kbug819 opened the pull request:
## Implements/Fixes: title-here
 **(testing/exploration only)**

Notes:
#70
Testing/exploration done for many-to-many relationships:

1. Created Preorder_test table
2. Created Preorder_testItems table
3. Created 3 new endpoints:
 - GET All preorders based on customer ID - http://127.0.0.1:8000/customers/1/preorder_test/
 - POST preorder based on customer ID - http://127.0.0.1:8000/customers/1/preorder_test/
 - GET All preorders that have an associated item with specific vendor - http://127.0.0.1:8000/vendors/1/preorder_test/
4. Created new serializers to handle returns
5. Created Views 

Sample endpoint return:
[{
        "id": 2,
        "customer": 1,
        "ready": true,
        "packed": false,
        "fulfilled": false,
        "items": [
            {
                "item_id": 3,
                "item_name": "Sheep",
                "vendor_id": 1,
                "quantity": 2
            },
            {
                "item_id": 6,
                "item_name": "Tomato",
                "vendor_id": 2,
                "quantity": 4
            }
        ]
    },
    {
        "id": 5,
        "customer": 1,
        "ready": true,
        "packed": false,
        "fulfilled": false,
        "items": [
            {
                "item_id": 3,
                "item_name": "Sheep",
                "vendor_id": 1,
                "quantity": 10
            },
            {
                "item_id": 6,
                "item_name": "Tomato",
                "vendor_id": 2,
                "quantity": 5
            }
        ]
    }]

## This closes issue:
 (add issue number here) 


## Necessary checkmarks:
- [ ] All Tests are Passing
- [X] The code will run locally


## Type of change
- [X] New feature
- [ ] Bug Fix


## Check the correct boxes
- [X] This broke nothing
- [ ] This broke some stuff
- [ ] This broke everything


## Testing Changes
- [ ] No Tests have been changed
- [ ] Some Tests have been changed
- [ ] All of the Tests have been changed(Please describe what in the world happened)


## Checklist:
- [ ] My code has no unused/commented out code
- [ ] I have reviewed my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have fully tested my code


TY!

Antoine-Aube commented on market_nextdoor_api/migrations/0004_preorder_test_preorder_testitem.py:
```diff
@@ -0,0 +1,39 @@
+                ('date_created', models.DateTimeField(auto_now_add=True)),
+                ('updated_at', models.DateTimeField(auto_now=True)),
+                ('item', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='market_nextdoor_api.item')),
+                ('preorder', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='market_nextdoor_api.preorder_test')),
```
Curious about this delete function. Looks similar to dependent destroy in rails?

kbug819 replied:
Right, very similar, if the associated preorder is deleted, the associated items will also be deleted.

Antoine-Aube commented on market_nextdoor_api/models.py:
```diff
@@ -68,10 +68,24 @@ class Preorder(models.Model):
-#     quantity = models.IntegerField(default=1)
-#     date_created = models.DateTimeField(auto_now_add=True)
-#     updated_at = models.DateTimeField(auto_now=True)
+class Preorder_testItem(models.Model):
```
Got it, so essentially just the joins table here. Nothing too crazy! Still getting used to this syntax for sure.

Antoine-Aube commented on market_nextdoor_api/serializers.py:
```diff
@@ -36,4 +36,24 @@ class Meta:
+  
+
+# ManytoMany Test Serializer
+class Preorder_testItemSerializer(serializers.ModelSerializer):
```
Is there a reason why we would want to serialize this? Are we ever going to return the preorder_item object to the Front end or will they just query it for a list of all the items on the preorder? I could definitely be overthinking this aspect.

Antoine-Aube replied:
oh I think I can maybe answer my own question from looking at the below, serializing isn't just for the response, but for object creation in the db, correct?

kbug819 replied:
We aren't really creating in the serializer, but I built the Preorder_testItem serializer because the items are nested in the Preorder_test serializer. Preorder_testSerializer calls the item serializer and accesses Items within the fields.

Antoine-Aube commented on market_nextdoor_api/views/preorder_views.py:
```diff
@@ -62,4 +62,38 @@ def update_preorder(preorder, data):
+    return Response(status=status.HTTP_404_NOT_FOUND)
+
+  if request.method == 'GET':
+    return get_preorder_test_list(request, check_customer)
```
is the check_customer argument representing customer id that will be passed from FE?

kbug819 replied:
In this GET request, FE will be sending the customer id through the URL: '/customers/1/preorders/' - check customer finds the Customer object based on the ID passed in the request:
"try:
    check_customer = Customer.objects.get(pk=customer_id)"

Antoine-Aube replied:
I guess I just saw that as an argument and not a method. So the argument is a method? very interesting.

Antoine-Aube commented on market_nextdoor_api/views/preorder_views.py:
```diff
@@ -62,4 +62,38 @@ def update_preorder(preorder, data):
+        quantity=item['quantity']
+      )
+    return Response(serializer.data, status=status.HTTP_201_CREATED)
+  return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
would an 'else' statement here make sense? It seems as though this method will return a bad request as an alternative to everything else, which isn't technically incorrect, but could make it a bit more readable. Once again, newbie here making suggestions so probably wrong.

kbug819 replied:
Agreed, when we build out functionality for this on main, will definitely need to add further logic to avoid bad requests.

Antoine-Aube commented on market_nextdoor_api/views/vendor_views.py:
```diff
@@ -108,4 +108,22 @@ def update_preorder(preorder, data):
+    return get_preorder_test_list(request, check_vendor)
+
+def get_preorder_test_list(request, check_vendor):
+  preorder_tests = Preorder_test.objects.filter(items__vendor=check_vendor).distinct()
```
is the .distinct operation here built into django? ORM at all or just a method that comes as a part of library?

kbug819 replied:
.distint() is part of the Django Framework and is part of the Django ORM.

Antoine-Aube commented on market_nextdoor_api/models.py:
```diff
@@ -68,10 +68,24 @@ class Preorder(models.Model):
+#ManytoMany Test Models
+class Preorder_test(models.Model):
+    customer = models.ForeignKey(Customer, on_delete=models.CASCADE, null=False)
+    items = models.ManyToManyField(Item, through='Preorder_testItem')
```
this is looking familiar with the use of "Through" in terms of has many items through preorder_items.

Antoine-Aube reviewed (approved):
Overall looks good to me, although I have to admit this brought up more questions for me about database querying, adding objects through serializers, etc. Would love to poke around a bit more with you both tomorrow to answer some questions!

1. Serializer = way to write data to the db? Does that hand in hand with commands like .create, .save, etc? 

2. In the tests, which I assume are for the old preorder classes, what does "deserialize" do?

DylanScotty commented on market_nextdoor_api/migrations/0004_preorder_test_preorder_testitem.py:
```diff
@@ -0,0 +1,39 @@
+        migrations.CreateModel(
+            name='Preorder_test',
+            fields=[
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
```
This looks great! I have a question here about BigAutoField for the id, curious what this is or what it is doing.

kbug819 replied:
I think that's the syntax for the migration to automatically add an id to each entry?

kbug819 commented:
Updated the following after team meeting:

1. Naming conventions update
- preorder_test changed to **preorder2**
- preorder_testItem changed to **preorder2Item**
2. Added further logic for vendor preorders and customer preorders
3. Rearranged the files to make things easier to find
4. Testing completed for CRUD and edge/sad path cases

EVanGorkom commented:
This looks good. While I'm working on the Markets many-to-many, I was thinking about implementing versioning if that sounds good to you. As far as the CRUD functions, I'm fine with customers updating their own preorders or deleting their orders personally. Like if that want to cancel an order then they should be able to do so etc. Maybe flag a flash message "are you sure you want to delete this preorder", but that's a front end thing.

kbug819 commented:
@EVanGorkom, versioning sounds good to me. Was also doing some research on how to implement last night. Will do some further research today. Keep me updated with what you come up with.

EVanGorkom reviewed (approved):
Thank you for uncommenting the CRUD functions for customer preorders! I'll look into potentially versioning our endpoints so that we don't have to have commented out code/ different names etc.
